### PR TITLE
[GPT-OSS] Add support for batch=128 on WH-Galaxy

### PR DESCRIPTION
**Pull request by handrewsTT**
### Ticket
[#29623](https://github.com/tenstorrent/tt-metal/issues/29623)

### What's changed
1) Implement EP=32 on 32 device mesh (4 experts per device for 120b, 1 for 20b)

2) Add support for user (row) sharded batch=128 implementation of gpt-oss on WH GLX. 

3) Move accuracy tests to demo workflows

4) Add more unit tests

### Checklist
- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:handrews/gpt-oss-b128)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:handrews/gpt-oss-b128)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:handrews/gpt-oss-b128)
- [ ] [(T3K) T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml?query=branch:handrews/gpt-oss-b128)
- [ ] [(T3K) T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml?query=branch:handrews/gpt-oss-b128)
- [ ] [(Galaxy) unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-unit-tests.yaml?query=branch:handrews/gpt-oss-b128)
- [ ] [(Galaxy) demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml?query=branch:handrews/gpt-oss-b128)

### CODE REVIEW

**Comment by handrewsTT** [comment]
@copilot review this PR and summarise the changes along with a recommendation for merge or request changes

**Review by sraizada-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/ccl/all_to_all_dispatch/device/all_to_all_dispatch_device_operation.cpp [hunk 1]
```diff
@@ -27,7 +27,7 @@ void AllToAllDispatchDeviceOperation::validate_on_program_cache_miss(
     TT_FATAL(indices_tensor.layout() == tt::tt_metal::Layout::ROW_MAJOR, "Indices tensor must be in row major layout");
 
     TT_FATAL(input_tensor.dtype() == tt::tt_metal::DataType::BFLOAT16, "Input tensor must be bfloat16");
-    TT_FATAL(indices_tensor.dtype() == tt::tt_metal::DataType::UINT16, "Indices tensor must be uint32");
+    TT_FATAL(indices_tensor.dtype() == tt::tt_metal::DataType::UINT16, "Indices tensor must be uint16");
```
Comment (jvegaTT) [inline ttnn/cpp/ttnn/operations/ccl/all_to_all_dispatch/device/all_to_all_dispatch_device_operation.cpp:31]: Good catch! Sorry for the mistype

**Review by jvegaTT — APPROVED** [review]

**Comment by yieldthought** [comment]
Codex review:

The decode path now hard-codes a padded batch dimension of 32 and the row-sharded prefill path hard-codes a 128-row page table; both introduce shape mismatches that will break or misroute workloads when batch sizes deviate from those constants.

Full review comments:

- [P1] Pad decode reshape using the actual batch size -- [models/demos/gpt_oss/tt/attention/decode.py:156](https://github.com/tenstorrent/tt-metal/blob/cbec657bbd7fdcf177d4cdfa8f0391ccda799168/models/demos/gpt_oss/tt/attention/decode.py#L156)
  When batch_size exceeds 32 (e.g., the row-sharded 128-batch decode path), the reshape uses a fixed padded shape of (1, 1, 32, hidden_size), which is smaller than the logical shape (1, 1, batch_size, hidden_size); ttnn.reshape will either assert or truncate/garble the output in that case. The padded dimension should be derived from batch_size (e.g., nearest_32(batch_size)) or the existing tt_out padded shape rather than hard-coding 32.

- [P2] Size row-sharded page tables from max batch size -- [models/demos/gpt_oss/tt/generator.py:182](https://github.com/tenstorrent/tt-metal/blob/cbec657bbd7fdcf177d4cdfa8f0391ccda799168/models/demos/gpt_oss/tt/generator.py#L182)
  get_prefill_page_table always pads to 128 rows. If max_batch_size is not 128 (e.g., vLLM configs that use 64 or 96), the page table batch dimension won't match the KV cache allocation, which can make paged_fill_cache fail or write into the wrong slots. The padded table should use the configured max batch (e.g., self.model_args[0].max_batch_size or kv_cache[0].shape[0]) instead of a constant.

**Review by yieldthought — APPROVED** [review]
Take a look at the codex comments, probably worth guarding against these failures

**Review by roseli-TT — APPROVED** [review]

File: models/demos/gpt_oss/tt/generator.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+import torch
+from loguru import logger
+
+import ttnn
+from models.tt_transformers.tt.common import get_block_size, get_padded_prefill_len, num_blocks_in_seq
+from models.tt_transformers.tt.generator import Generator
+
+
+class GPTOSSRowShardedGenerator(Generator):
```
Comment (skhorasganiTT) [inline models/demos/gpt_oss/tt/generator.py]: I do not think this forked Generator should exist and I am concerned about the way that people have started using empty_slots by treating it as global user ids:
- The prefill_forward_text function in this Generator is nearly the same as the one from TTT, except it's accessing the page table differently and it's also using TTT code from an older main, already illustrating the harm from forking before this has even landed
- Diverging Generators will cause great difficulty with adding Generator-level features to different models such as chunked prefill, prefix caching, trace, etc.
- empty_slots is an artifact of the V0 DP workaround in vLLM, and it was never intended to be treated as global user ids or kept long term. It has been kept in V1 just for backwards compatibility and is meant to be replaced with a different input which only states the DP cache group corresponding to each user. No TT model should be using these as "user IDs" as they are not real global user ids and may change across decode iterations in V1. Currently it's being used by Llama70B-GLX (and subsequently DeepSeek, and now this model) to re-shuffle the page table into a certain order which has the user at a specific index, so that paged_fill_cache can then access that one index for the appropriate values in the page table. However this is redundant since the index in dim 0 of the page table (also `idx` in for idx, user_id in enumerate(empty_slots)) already gives the appropriate page table row for the current user (see TT-Transformers generator here - https://github.com/tenstorrent/tt-metal/blob/6346e1bb728e9ffaf308d8ebf98f438a3e3c82f0/models/tt_transformers/tt/generator.py#L332). 

cc @sraizada-tt

**Review by skhorasganiTT — CHANGES_REQUESTED** [review]

File: models/demos/gpt_oss/tt/generator.py (hunk 2 above)
Comment (sraizada-tt) [reply models/demos/gpt_oss/tt/generator.py]: done

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -281,8 +281,7 @@ def prefill_forward_text(
             # Only paged attention is supported for prefill
             enable_trace = False
 
-        # we need this here becuase of tt-metal tests
-        self.warmup_model_prefill(kv_cache, enable_trace)
+        # self.warmup_model_prefill(kv_cache, enable_trace)
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: accident?

**Review by skhorasganiTT — CHANGES_REQUESTED** [review]

File: models/tt_transformers/tt/generator.py (hunk 3 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: no, it's just annoying :)

**Review by skhorasganiTT — APPROVED** [review]
